### PR TITLE
start updating gridop and its links with postp

### PR DESCRIPTION
**Pull request by apatlpo**
The scope of this PR is to update and improve gridop, in particular its links with postp by for example adding wrappers to gridop methods in postp run objects.

### CODE REVIEW

**Comment by apatlpo** [comment]
@slgentil @NoeLahaye 
This PR is not ready for review yet.

**Comment by apatlpo** [comment]
This is moving along allright.

I would need input from you @NoeLahaye though:
I am currently looking at the vertical mode code and I believe it could be much improved by wrapping all methods into a single code. I can give it a shot but would need to know a couple of things:

I have the feeling vmodes.py is up to date. Do you confirm ?

I don't see any code for modal projection except in the notebook proj_modes.ipynb. Is this up to date?

Also is get_pressure in gridop.py up to date?

thanks

**Comment by NoeLahaye** [comment]
Answersand questions for @apatlpo below

> This is moving along allright.
> 
> I would need input from you @NoeLahaye though:
> I am currently looking at the vertical mode code and I believe it could be much improved by wrapping all methods into a single code. I can give it a shot but would need to know a couple of things:
> 
> I have the feeling vmodes.py is up to date. Do you confirm ?
Yes indeed, vmodes.py is up-to-date. A quick reminder : the first function is a wrapper that should be called through apply_ufunc with dask arrays as arguments -- this is the only function that is being used explicitly; compute_vmodes is a wrapper that takes ND numpy arrays and call the actual solver: compute_vmodes_1D, which works with 1D (z) columns, in numpy.
> 
> I don't see any code for modal projection except in the notebook proj_modes.ipynb. Is this up to date?
in tutorials/dev: DiagDynCSW_proj_v1.ipynb and DiagDynCSW_proj_white.ipynb are working version of notebooks that do the projection and compute the projected equations terms. I will update proj_modes.ipynb in the tutorials to provide a minimal working example this week.
> 
> Also is get_pressure in gridop.py up to date?
Not quite: the version with zr specified (i.e. computing pressure on rho points) is OK, but I have an issue with grid.cumsum for the case where the pressure is computed on w-points (which is probably not useful but I still wanted to have this).
I need to cumsum grom in a reverse order (i.e. from the top to bottom) but grid.cumsum seems to ignore the sortby(s_rho, ascending=False). Any clue as of why ?
> 
> thanks

**Comment by apatlpo** [comment]
ok, thanks @NoeLahaye 

vmodes is outside of the scope of this PR.
I laid out the structure of what the class should look like.
It would be great if you could do another PR implementing the vmode class. Do you think you could do that soon?

Before that though, I'd like to go through with the present PR.
Could you please review it?
To do that you have to click on `Files changed` on the present page and start the review process.

Here is a summary of the modifications:

- add wrappers in `postp.Run` objects to gridop functions
- delete all functions of the type `rho2u`, etc and replace them with `x2u` (any point to u)
- work on `get_z`, in particular spatial dimension order
- work on `N2Profile`, called now `get_N2`
- `get_p`, `get_uv...`, `get_pv` have been moved around but not modified
- backward compatibility should be maintained and improved (but this is hard)

I'll get back to you soon about grid.cumsum

**Comment by NoeLahaye** [comment]
I don't know where to put this question, but could we read the value of g from the simulation? That would be better than hardcoding it everywhere...

File: crocosi/gridop.py [hunk 1]
```diff
@@ -1,179 +1,142 @@
...
     if hgrid in ['u','v']:
-        funtr = eval("rho2"+hgrid)
-        if zeta is None:
-            _zeta = funtr(_zeta, ds)
-        _h = funtr(_h, ds)
+        _h = x2x(_h, xgrid, hgrid)
+        _zeta = x2x(_h, xgrid, hgrid)
+    
+    # align datasets (zeta may contain a slice along one dimension for example)
+    _h, _zeta  = xr.align(_h, _zeta, join='inner')
```
Comment (NoeLahaye) [inline crocosi/gridop.py]: why do we need to align? Isn't it automatic when doing the calculation?

**Comment by apatlpo** [comment]
regarding your question about cumsum. It seems to be working for me, see
this [gist](https://gist.github.com/apatlpo/f921849ce092be2d8baa6fe1fe94f131)

regarding `g`, yes you're right, it's kind of akward at the moment. 
Let's write a default value once in postp (arbitrary choice) and import this value everywhere else.

File: crocosi/gridop.py (hunk 1 above)
Comment (NoeLahaye) [reply crocosi/gridop.py]: J'ai fait n'importe quoi, je maitrise pas encore bien, le commentaire sur vmodes ne devrait pas apparaitre dans cette review...

Comment (apatlpo) [reply crocosi/gridop.py]: imagine zeta is a y slice at one x location and that h was not provided.
In that case, h will be loaded from grid and hence 2D. 
If we do not align with the 'inner' option, zeta will eventually be broadcasted which I don't think is desirable.

**Comment by apatlpo** [comment]
ok, thanks, your latest script will be broken unfortunately, I'm pretty sure.
But it should be relatively straightforward to make them work again.

You could potentially load this version of the code but I suggest we merge and you try to recover from the changes. You can always go back to the current version of master if this is too much of a problem while we fix things if necessary. Agree?

**Comment by NoeLahaye** [comment]
OK!

File: crocosi/gridop.py [hunk 2]
```diff
@@ -1,179 +1,142 @@
...
 
-    # swith horizontal grid if needed (should we use grid.interp instead?)
+    # switch horizontal grid if needed
     if hgrid in ['u','v']:
-        funtr = eval("rho2"+hgrid)
-        if zeta is None:
-            _zeta = funtr(_zeta, ds)
-        _h = funtr(_h, ds)
+        _h = x2x(_h, xgrid, hgrid)
+        _zeta = x2x(_h, xgrid, hgrid)
```
Comment (NoeLahaye) [inline crocosi/gridop.py]:
```suggestion
        _zeta = x2x(_zeta, xgrid, hgrid)
```

File: crocosi/gridop.py (hunk 1 above)
Comment (NoeLahaye) [reply crocosi/gridop.py]: OK, I see. Then we should avoid calling get_z with zeta=0, right?

File: crocosi/gridop.py (hunk 2 above)
Comment (apatlpo) [reply crocosi/gridop.py]: good catch

**Comment by apatlpo** [comment]
all right let's merge this

**Pull request merged**